plugins/shadow-block-converter/src/shadow_block_converter.js

```diff
@@ -0,0 +1,194 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Colour input field with HSV sliders.
+ */
+
+import * as Blockly from 'blockly/core';
+
+/**
+ * A new blockly event class specifically for recording changes to the shadow
+ * state of a block. This implementation is similar to and could be merged with
+ * the implementation of Blockly.Events.BlockChange in Blockly core code.
+ * @extends {Blockly.Events.BlockBase}
+ */
+export class BlockShadowChange extends Blockly.Events.BlockBase {
+  /**
+   * The name of the event type for broadcast and listening purposes.
+   * @type {string}
+   */
+  static EVENT_TYPE = 'block_shadow_change';
+
+  /**
+   * The constructor for a new BlockShadowChange event.
+   * @param {Blockly.Block=} block The changed block. Undefined for a blank
+   *   event.
+   * @param {boolean=} oldValue Previous value of shadow state.
+   * @param {boolean=} newValue New value of shadow state.
+   */
+  constructor(block, oldValue, newValue) {
+    super(block);
+
+    /**
+     * Type of this event.
+     * @type {string}
+     */
+    this.type = BlockShadowChange.EVENT_TYPE;
+
+    if (!block) {
+      return; // Blank event to be populated by fromJson.
+    }
+    this.oldValue = typeof oldValue === 'undefined' ? '' : oldValue;
+    this.newValue = typeof newValue === 'undefined' ? '' : newValue;
+  }
+
+  /**
+   * Encode the event as JSON.
+   * @return {!Object} JSON representation.
+   * @override
+   */
+  toJson() {
+    const json = super.toJson();
+    json['oldValue'] = this.oldValue;
+    json['newValue'] = this.newValue;
+    return json;
+  }
+
+  /**
+   * Decode the JSON event.
+   * @param {!Object} json JSON representation.
+   * @override
+   */
+  fromJson(json) {
+    super.fromJson(json);
+    this.oldValue = json['oldValue'];
+    this.newValue = json['newValue'];
+  }
+
+  /**
+   * Does this event record any change of state?
+   * @return {boolean} False if something changed.
+   * @override
+   */
+  isNull() {
+    return this.oldValue === this.newValue;
+  }
+
+  /**
+   * Run a change event.
+   * @param {boolean} forward True if run forward, false if run backward (undo).
+   * @override
+   */
+  run(forward) {
+    const workspace = this.getEventWorkspace_();
+    const block = workspace.getBlockById(this.blockId);
+    if (!block) {
+      console.warn('Can\'t change non-existent block: ' + this.blockId);
+      return;
+    }
+
+    const value = forward ? this.newValue : this.oldValue;
+    block.setShadow(!!value);
```

Comment:
What happens if `block` has children and you're making `block` into a shadow? E.g. the addition block here:
![image](https://user-images.githubusercontent.com/13686399/192909983-5f3fc946-d57b-4796-9f2d-985e80a7a1f6.png)

Reply:
Currently, nothing. There's nothing that's listening for shadow change events, and shadow change events themselves only affect the one block they're associated with. Shadow change events are only created in response to new block changed events, in which case they reify the block and any parents if necessary, and ignore children of the changed block.